photon-ml/src/main/scala/com/linkedin/photon/ml/optimization/AbstractOptimizer.scala

```diff
@@ -1,132 +0,0 @@
-
-  protected var previousState: Option[OptimizerState] = None
-
-  protected var statesTracker: Option[OptimizationStatesTracker] = if (isTrackingState) {
-    Some(new OptimizationStatesTracker())
-  } else {
-    None
-  }
-
-  def isDone: Boolean = convergenceReason.nonEmpty
-
-  def convergenceReason: Option[ConvergenceReason] = {
-    if (initialState.isEmpty || currentState.isEmpty || previousState.isEmpty) {
-      None
-    } else if (currentState.get.iter >= maxNumIterations) {
-      Some(MaxIterations)
-    } else if (currentState.get.iter == previousState.get.iter) {
-      Some(ObjectiveNotImproving)
-    } else if (abs(currentState.get.value - previousState.get.value) <= tolerance * initialState.get.value) {
-      Some(FunctionValuesConverged)
-    } else if (norm(currentState.get.gradient, 2) <= tolerance * norm(initialState.get.gradient, 2)) {
-      Some(GradientConverged)
-    } else {
-      None
-    }
-  }
-
-  def getTolerance: Double = tolerance
-
-  def setTolerance(tol: Double) {
-    require(!(tol.isInfinite || tol.isNaN) && tol > 0.0)
-    tolerance = tol
-  }
-
-  def getMaximumIterations: Int = maxNumIterations
-
-  def setMaximumIterations(maxIter: Int) {
-    require(maxIter > 0)
-    maxNumIterations = maxIter
-  }
-
-  def getConstraintMap: Option[Map[Int, (Double, Double)]] = constraintMap
-
-  def setConstraintMap(constraints: Option[Map[Int, (Double, Double)]]) {
-    constraintMap = constraints
-  }
-
-  def setStateTrackingEnabled(enabled: Boolean) {
-    isTrackingState = enabled
-    statesTracker = if (isTrackingState) Some(new OptimizationStatesTracker()) else None
-  }
-
-  def getStateTracker: Option[OptimizationStatesTracker] = statesTracker
-
-  def getInitialState: Option[OptimizerState] = initialState
-
-  protected def setInitialState(state: Option[OptimizerState]): Unit = {
-    initialState = state
-  }
-
-  def getCurrentState: Option[OptimizerState] = currentState
-
-  protected def setCurrentState(state: Option[OptimizerState]): Unit = {
-    (state, statesTracker, currentState) match {
-      case (Some(s), Some(st), Some(ct)) =>
-        // Only tracks the state if it is different from the previous state (e.g., different objects)
-        if (s != ct) {
-          st.track(s)
-        }
-      case _ =>
-    }
-    currentState = state
-  }
-
-  protected def setConvergenceReason(): Unit = {
-    // Set the convergenceReason when the optimization is done
-    statesTracker match {
-      case Some(y) => y.convergenceReason = convergenceReason
-      case None =>
-    }
-  }
-
-  def getPreviousState: Option[OptimizerState] = previousState
-
-  protected def setPreviousState(state: Option[OptimizerState]): Unit = {
-    previousState = state
-  }
-
-  def stateTrackingEnabled: Boolean = isTrackingState
-
-  protected def clearOptimizationStatesTracker(): Unit = {
-    statesTracker match {
-      case Some(x) => x.clear()
-      case None =>
-    }
-  }
-}
```

Comment:
When Brendan designed this class, the motivation was to let ```AbstractOptimizer``` to handle the bookkeeping of the optimizer's inner states, because such bookkeeping was independent of the optimizer's core logic: optimization. And the plan was that in the future if we find a better way to handle the inner states, only components in ```AbstractOptimizer``` needs to be changed with the optimizer remaining intact.